### PR TITLE
FAQ algorithm

### PR DESCRIPTION
**Pull request by aurorarossi**
Hello!
In this PR I implemented the graph alignment algorithm 1 of [this paper](https://arxiv.org/pdf/2111.05366.pdf).

### CODE REVIEW

File: test/GOATalgorithm.jl [hunk 1]
```diff
@@ -0,0 +1,97 @@
+function check_doubly_stochastic(D::AbstractMatrix{U}) where {U<:Real}
+    n=size(D)[1]
+    for i in 1:n
+        @test 0.99<sum(D[i,:])<1.01 #check sum of elements in a row == 1
```
Comment (natema) [inline test/GOATalgorithm.jl]: What about using `@test sum(D[i,:]) ≈ 1` instead?

File: Project.toml [hunk 2]
```diff
@@ -4,10 +4,14 @@ authors = ["Guillaume Dalle <22795598+gdalle@users.noreply.github.com> and contr
 version = "0.1.0"
 
 [deps]
+GLPK = "60bf3e95-4087-53dc-ae20-288a0d20c6a6"
 Graphs = "86223c79-3864-5bf0-83f7-82e725a168b6"
+HiGHS = "87dc4568-4c63-4d18-b0c0-bb2238e4078b"
```
Comment (natema) [inline Project.toml:9]: I think it is better not to include the solvers by default.

File: src/GOATalgorithm.jl [hunk 3]
```diff
@@ -0,0 +1,152 @@
+#flat doubly stochastic matrix, starting point
+"""
+    _flat_doubly_stochastic(n)
+
+Given the dimension 'n', return the flat doubly stochastic matrix with number of rows and columns equals to 'n'.
+The flat doubly stochastic matrix is the barycenter of doubly stochastic matrices.
```
Comment (natema) [inline src/GOATalgorithm.jl]: It would be better to just write here in $\LaTeX$ that it is the matrix $\frac 1n \mathbb 1^T \mathbb 1$.

File: src/GOATalgorithm.jl [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+"""
+_flat_doubly_stochastic(n::Int64)=ones(n) * ones(n)' / n
+
+#gradient of the objective function
+"""
+    _gradient(A,B,P_i)
+
+Given the adjacency matrices A and B, return the gradient evaluated in P_i of the function f(P)=-tr(A*P*B'*P').
```
Comment (natema) [inline src/GOATalgorithm.jl]: Here and in the rest of the docstrings also, maybe it looks nicer wrapping the expressions in LaTeX delimiters.

File: src/GOATalgorithm.jl [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+function _check_convergence(gradient::AbstractMatrix{U},P_i::AbstractMatrix{V},P_new::AbstractMatrix{T};tol::Float64) where {U<:Real,V<:Real,T<:Real}
+    if norm(gradient)*(norm(P_i-P_new))<tol
+        return true
+    else
+        return false
+    end
+end
+
+"""
+    goat(A,B;optimizer,max_iter=30,tol=0.1,init+=_flat_doubly_stochastic(size(A)[1]))
```
Comment (natema) [inline src/GOATalgorithm.jl]: @aurorarossi I think it is better to call it FAQ in the end. 
In either case, it's better to specify below that 
- it is using optimal transport, unlike the original 2015 paper,
- it is not using the "lighting" version of optimal transport introduced in the GOAT paper.

File: test/GOATalgorithm.jl [hunk 6]
```diff
@@ -0,0 +1,97 @@
+function check_doubly_stochastic(D::AbstractMatrix{U}) where {U<:Real}
+    n=size(D)[1]
+    for i in 1:n
+        @test sum(D[i,:]) ≈ 1  #check sum of elements in a row == 1
```
Comment (natema) [inline test/GOATalgorithm.jl]: Extra space to fix.

**Review by natema — CHANGES_REQUESTED** [review]

File: test/GOATalgorithm.jl (hunk 6 above)
Comment (natema) [reply test/GOATalgorithm.jl]: To be clear, I mean that here it should be 
```
        @test sum(D[i,:]) ≈ 1 #check sum of elements in a row == 1
```

Comment (natema) [reply test/GOATalgorithm.jl]: In general, as we checked, it is better to leave a space also after the `#`.

**Review by natema — APPROVED** [review]
I checked the generated docs with `LiveServer` and the $\LaTeX$ is correctly displayed.

**Comment by aurorarossi** [comment]
Schermata 2023-01-06 alle 00 43 32</img>
Comparison of the objective function values computed with this Julia implementation, a Python [implementation](https://github.com/microsoft/graspologic) and the correct values of 40 instances of the [QAP library](https://www.opt.math.tugraz.at/qaplib/inst.html).

File: Project.toml [hunk 7]
```diff
@@ -8,8 +8,10 @@ FillArrays = "1a297f60-69ca-5386-bcde-b61e274b549b"
 Graphs = "86223c79-3864-5bf0-83f7-82e725a168b6"
 HiGHS = "87dc4568-4c63-4d18-b0c0-bb2238e4078b"
 JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 SimpleTraits = "699a6c99-e7fa-54fc-8d76-47d257e15c1d"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Comment (gdalle) [inline Project.toml]: Why is this needed? It shouldn't be a dependency of the package itself

File: src/GraphsOptim.jl [hunk 8]
```diff
@@ -16,9 +16,12 @@ using JuMP: set_silent, optimize!, termination_status, value
 using JuMP: @variable, @constraint, @objective
 using SimpleTraits: SimpleTraits, @traitfn
 using SparseArrays: sparse
+using LinearAlgebra: norm, tr
 
 export minimum_cost_flow!, minimum_cost_flow
+export faq
```
Comment (gdalle) [inline src/GraphsOptim.jl:22]: Could we give it a more explicit name, with a dispatch mechanism for algorithm choice?
See https://github.com/JuliaGraphs/Graphs.jl/issues/264

File: src/alignment_algorithms.jl [hunk 9]
```diff
@@ -0,0 +1,148 @@
+# flat doubly stochastic matrix, starting point
+"""
+    _flat_doubly_stochastic(n)
+
+Given the dimension ``n``, return the flat doubly stochastic matrix ``J=\\frac{1}{n}\\mathbb{1}\\mathbb{1}^T``.
+The flat doubly stochastic matrix is the barycenter of doubly stochastic matrices.
+"""
+_flat_doubly_stochastic(n::Int64) = ones(n) * ones(n)' / n
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Might we need type genericity here? At the moment this returns a matrix of `Float64`. Also, the `n::Int64` can be replaced with `n::Integer`

File: src/alignment_algorithms.jl [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+# gradient of the objective function
+"""
+    _gradient(A,B,P_i)
+
+Given the adjacency matrices ``A`` and ``B``, return the gradient evaluated in ``P_i`` of the function ``f(P)=-\\operatorname{trace}(APB^TP^T)``.
+"""
+function _gradient(
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: More informative function name?

File: src/alignment_algorithms.jl [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    return -A * P_i * B' - A' * P_i * B
+end
+
+# norm
+"""
+    _distance(A,B,P)
+
+Given the adjacency matrices ``A`` and ``B`` and the permutation matrix ``P``, compute the distance between the permuted graphs.
+"""
+function _distance(
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: More informative function name?

File: src/alignment_algorithms.jl [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    end
+    return P
+end
+
+"""
+    _step_size(A,B,P_i,Q_i)
+
+Given ``A,B,P_i`` and the direction matrix ``Q_i``, return the step size of the gradient descent method.
+"""
+function _step_size(
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: More informative function name?

File: src/alignment_algorithms.jl [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+Given the dimension ``n``, return the flat doubly stochastic matrix ``J=\\frac{1}{n}\\mathbb{1}\\mathbb{1}^T``.
+The flat doubly stochastic matrix is the barycenter of doubly stochastic matrices.
+"""
+_flat_doubly_stochastic(n::Int64) = ones(n) * ones(n)' / n
+
+# gradient of the objective function
+"""
+    _gradient(A,B,P_i)
+
+Given the adjacency matrices ``A`` and ``B``, return the gradient evaluated in ``P_i`` of the function ``f(P)=-\\operatorname{trace}(APB^TP^T)``.
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Typo in the formula

File: src/alignment_algorithms.jl [hunk 14]
```diff
@@ -0,0 +1,148 @@
+# flat doubly stochastic matrix, starting point
+"""
+    _flat_doubly_stochastic(n)
+
+Given the dimension ``n``, return the flat doubly stochastic matrix ``J=\\frac{1}{n}\\mathbb{1}\\mathbb{1}^T``.
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: You can avoid double backslashs by using `@doc raw"""`
https://documenter.juliadocs.org/stable/man/latex/

File: src/alignment_algorithms.jl [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+# norm
+"""
+    _distance(A,B,P)
+
+Given the adjacency matrices ``A`` and ``B`` and the permutation matrix ``P``, compute the distance between the permuted graphs.
+"""
+function _distance(
+    A::AbstractMatrix{U}, B::AbstractMatrix{V}, P::AbstractMatrix{T}
+) where {U<:Real,V<:Real,T<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we need all the type parameters?

File: src/alignment_algorithms.jl [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    return norm(A * P - P * B)
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+# gradient of the objective function
+"""
+    _gradient(A,B,P_i)
+
+Given the adjacency matrices ``A`` and ``B``, return the gradient evaluated in ``P_i`` of the function ``f(P)=-\\operatorname{trace}(APB^TP^T)``.
+"""
+function _gradient(
+    A::AbstractMatrix{U}, B::AbstractMatrix{V}, P_i::AbstractMatrix{T}
+) where {U<:Real,V<:Real,T<:Real}
+    return -A * P_i * B' - A' * P_i * B
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+function _solve_assignment_problem(data::AbstractMatrix{U}; optimizer) where {U<:Real}
+    M, N = size(data)
+    model = Model(optimizer)
+    set_silent(model)
+    @variable(model, x[m in 1:M, n in 1:N] >= zero(U), integer = true)
+    @objective(model, Max, sum(data[m, n] * x[m, n] for n in 1:N, m in 1:M),)
+    @constraint(model, [m in 1:M], sum(x[m, :]) == one(U))
+    @constraint(model, [n in 1:N], sum(x[:, n]) == one(U))
+    optimize!(model)
+    P = zeros(M, N)
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This is typed as `Float64`, whereas we presumably want a `Bool` or `Int`

File: src/alignment_algorithms.jl [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+Given ``A,B,P_i`` and the direction matrix ``Q_i``, return the step size of the gradient descent method.
+"""
+function _step_size(
+    A::AbstractMatrix{U},
+    B::AbstractMatrix{V},
+    P_i::AbstractMatrix{T},
+    Q_i::AbstractMatrix{W},
+) where {U<:Real,V<:Real,T<:Real,W<:Real}
+    R = P_i - Q_i
+    a = -tr(A * R * B' * R')
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    b = -tr(A * Q_i * B' * R' + A * R * B' * Q_i')
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    _step_size(A,B,P_i,Q_i)
+
+Given ``A,B,P_i`` and the direction matrix ``Q_i``, return the step size of the gradient descent method.
+"""
+function _step_size(
+    A::AbstractMatrix{U},
+    B::AbstractMatrix{V},
+    P_i::AbstractMatrix{T},
+    Q_i::AbstractMatrix{W},
+) where {U<:Real,V<:Real,T<:Real,W<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we need all the type parameters?

File: src/alignment_algorithms.jl [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+    R = P_i - Q_i
+    a = -tr(A * R * B' * R')
+    b = -tr(A * Q_i * B' * R' + A * R * B' * Q_i')
+    if a > 0.0
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we risk numerical instability? Maybe use `a > eps(a)`?

File: src/alignment_algorithms.jl [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+        α = min(max(-0.5 * b / a, 0.0), 1.0)
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: `-0.5 * b / a` is less type-generic than `b / (2a)`

Comment (gdalle) [reply src/alignment_algorithms.jl]: same goes for `min(x, 1.0)` versus `min(x, one(x))`

Comment (gdalle) [reply src/alignment_algorithms.jl]: of course maybe it doesn't matter

File: src/alignment_algorithms.jl [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+end
+
+"""
+    _update_P(P_i,Q_i,α)
+
+Given the previous matrix, the direction and the step size, return the updated matrix.
+"""
+function _update_P(
+    P_i::AbstractMatrix{U}, Q_i::AbstractMatrix{V}, α::Float64
+) where {U<:Real,V<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we need all the type parameters?

File: src/alignment_algorithms.jl [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    return α
+end
+
+"""
+    _update_P(P_i,Q_i,α)
+
+Given the previous matrix, the direction and the step size, return the updated matrix.
+"""
+function _update_P(
+    P_i::AbstractMatrix{U}, Q_i::AbstractMatrix{V}, α::Float64
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Does `α` have to be a `Float64`?

File: src/alignment_algorithms.jl [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+    P_i::AbstractMatrix{T},
+    Q_i::AbstractMatrix{W},
+) where {U<:Real,V<:Real,T<:Real,W<:Real}
+    R = P_i - Q_i
+    a = -tr(A * R * B' * R')
+    b = -tr(A * Q_i * B' * R' + A * R * B' * Q_i')
+    if a > 0.0
+        α = min(max(-0.5 * b / a, 0.0), 1.0)
+    else
+        if a + b > 0.0
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Numerical instability?

File: src/alignment_algorithms.jl [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+            α = 0.0
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Does `α` have to be a `Float64`?

File: src/alignment_algorithms.jl [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+
+"""
+    _update_P(P_i,Q_i,α)
+
+Given the previous matrix, the direction and the step size, return the updated matrix.
+"""
+function _update_P(
+    P_i::AbstractMatrix{U}, Q_i::AbstractMatrix{V}, α::Float64
+) where {U<:Real,V<:Real}
+    P_new = α * P_i + (1 - α) * Q_i
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+    return norm(A * P - P * B)
+end
+
+# solve assignment problem
+"""
+    _solve_assignment_problem(data; optimizer)
+
+Given a data matrix and a JuMP optimizer, return the integer solution of the assignment problem.
+"""
+function _solve_assignment_problem(data::AbstractMatrix{U}; optimizer) where {U<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: More informative argument name than `data`?

File: src/alignment_algorithms.jl [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+    set_silent(model)
+    @variable(model, x[m in 1:M, n in 1:N] >= zero(U), integer = true)
+    @objective(model, Max, sum(data[m, n] * x[m, n] for n in 1:N, m in 1:M),)
+    @constraint(model, [m in 1:M], sum(x[m, :]) == one(U))
+    @constraint(model, [n in 1:N], sum(x[:, n]) == one(U))
+    optimize!(model)
+    P = zeros(M, N)
+    for n in 1:N
+        for m in 1:M
+            P[m, n] = value(x[m, n])
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Can we just do `P = value.(x)`?

File: src/alignment_algorithms.jl [hunk 31]
```diff
@@ -0,0 +1,148 @@
...
+    _check_convergence(gradient,P_i,P_new;tol)
+
+Given the gradient and two consecutive step matrices, compute if the method converged.
+"""
+function _check_convergence(
+    gradient::AbstractMatrix{U},
+    P_i::AbstractMatrix{V},
+    P_new::AbstractMatrix{T};
+    tol::Float64,
+) where {U<:Real,V<:Real,T<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we need all the type parameters?

File: src/alignment_algorithms.jl [hunk 32]
```diff
@@ -0,0 +1,148 @@
...
+"""
+    _check_convergence(gradient,P_i,P_new;tol)
+
+Given the gradient and two consecutive step matrices, compute if the method converged.
+"""
+function _check_convergence(
+    gradient::AbstractMatrix{U},
+    P_i::AbstractMatrix{V},
+    P_new::AbstractMatrix{T};
+    tol::Float64,
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Does `tol` have to be a `Float64`?

File: src/alignment_algorithms.jl [hunk 33]
```diff
@@ -0,0 +1,148 @@
...
+    P_new = α * P_i + (1 - α) * Q_i
+    return P_new
+end
+
+"""
+    _check_convergence(gradient,P_i,P_new;tol)
+
+Given the gradient and two consecutive step matrices, compute if the method converged.
+"""
+function _check_convergence(
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: More informative function name?

File: src/alignment_algorithms.jl [hunk 34]
```diff
@@ -0,0 +1,148 @@
...
+    gradient::AbstractMatrix{U},
+    P_i::AbstractMatrix{V},
+    P_new::AbstractMatrix{T};
+    tol::Float64,
+) where {U<:Real,V<:Real,T<:Real}
+    if norm(gradient) * (norm(P_i - P_new)) < tol
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: This allocates more than needed (maybe it doesn't matter)

File: src/alignment_algorithms.jl [hunk 35]
```diff
@@ -0,0 +1,148 @@
...
+
+"""
+function faq(
+    A::AbstractMatrix{U},
+    B::AbstractMatrix{T};
+    optimizer,
+    max_iter::Int64=30,
+    tol::Float64=0.1,
+    init::AbstractMatrix{V}=_flat_doubly_stochastic(size(A)[1]),
+) where {U<:Real,V<:Real,T<:Real}
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Do we need all the type parameters?

File: src/alignment_algorithms.jl [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+    for _ in 1:max_iter
+        Q_i = _solve_assignment_problem(_gradient(A, B, P_i); optimizer=optimizer)
+        α_i = _step_size(A, B, P_i, Q_i)
+        P_new = _update_P(P_i, Q_i, α_i)
+        converged = _check_convergence(_gradient(A, B, P_i), P_i, P_new; tol=tol)
+        converged && break
+        P_i = P_new
+    end
+    P = _solve_assignment_problem(P_i; optimizer=optimizer)
+    return P, _distance(A, B, P), converged
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Document the return type (a tuple with ...)

File: test/alignment_algorithms.jl [hunk 37]
```diff
@@ -0,0 +1,79 @@
+function check_doubly_stochastic(D::AbstractMatrix{U}) where {U<:Real}
+    n = size(D)[1]
+    for i in 1:n
+        @test sum(D[i, :]) ≈ 1  # check sum of elements in a row == 1
```
Comment (gdalle) [inline test/alignment_algorithms.jl]: You can also use `sum(D; dims=1)`

Comment (gdalle) [inline test/alignment_algorithms.jl]: `one(U)` is more generic than `1`

File: test/alignment_algorithms.jl [hunk 38]
```diff
@@ -0,0 +1,79 @@
+function check_doubly_stochastic(D::AbstractMatrix{U}) where {U<:Real}
+    n = size(D)[1]
+    for i in 1:n
```
Comment (gdalle) [inline test/alignment_algorithms.jl]: Iterating with `for i in eachindex(D, 1)` or `for j in eachindex(D, j)` is better style

File: test/alignment_algorithms.jl [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+        @test sum(D[i, :]) ≈ 1  # check sum of elements in a row == 1
+        @test sum(D[:, i]) ≈ 1 # check sum of elements in a column == 1
+    end
+end
+
+function check_permutation_matrix(A::AbstractMatrix{U}) where {U<:Real}
+    n = size(A)[1]
+    for j in 1:n
```
Comment (gdalle) [inline test/alignment_algorithms.jl]: You can also use `eachindex` here

File: test/alignment_algorithms.jl [hunk 40]
```diff
@@ -0,0 +1,79 @@
...
+        for i in 1:n
+            @test (A[i, j] == one(U) || A[i, j] == zero(U)) # check that entries are 1 or 0
```
Comment (gdalle) [inline test/alignment_algorithms.jl]: You can also do `@test all(isbinary, A)` where you defined `isbinary`

File: test/alignment_algorithms.jl [hunk 41]
```diff
@@ -0,0 +1,79 @@
...
+            _, norm, converged = faq(A, B; optimizer=HiGHS.Optimizer)
+            @test norm == 0.0
+            @test converged
+        end
+
+        @testset "two path and complete graph" begin
+            A = adjacency_matrix(path_graph(3))
+            B = adjacency_matrix(complete_graph(3))
+            _, norm, converged = faq(A, B; optimizer=HiGHS.Optimizer)
+            @test norm == sqrt(2)
```
Comment (gdalle) [inline test/alignment_algorithms.jl]: Use `isapprox` (or the wiggly equal)

File: src/alignment_algorithms.jl [hunk 42]
```diff
@@ -0,0 +1,148 @@
...
+    _solve_assignment_problem(data; optimizer)
+
+Given a data matrix and a JuMP optimizer, return the integer solution of the assignment problem.
+"""
+function _solve_assignment_problem(data::AbstractMatrix{U}; optimizer) where {U<:Real}
+    M, N = size(data)
+    model = Model(optimizer)
+    set_silent(model)
+    @variable(model, x[m in 1:M, n in 1:N] >= zero(U), integer = true)
+    @objective(model, Max, sum(data[m, n] * x[m, n] for n in 1:N, m in 1:M),)
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: I wonder if `dot(data, x)` works here?

File: src/alignment_algorithms.jl [hunk 43]
```diff
@@ -0,0 +1,148 @@
...
+"""
+function faq(
+    A::AbstractMatrix{U},
+    B::AbstractMatrix{T};
+    optimizer,
+    max_iter::Int64=30,
+    tol::Float64=0.1,
+    init::AbstractMatrix{V}=_flat_doubly_stochastic(size(A)[1]),
+) where {U<:Real,V<:Real,T<:Real}
+    P_i = init
```
Comment (gdalle) [inline src/alignment_algorithms.jl]: Why name everything with `_i` ?

**Review by gdalle — DISMISSED** [review]
Hi @aurorarossi, nice PR!
My first review is mostly focused on style and good coding practices. Feel free to ignore some of the remarks, but I think writing type-generic and allocation-free code is a better paradigm in general.
Once you address these comments, I'll do a second pass on the actual content of the algorithm!

File: src/alignment_algorithms.jl (hunk 13 above)
Comment (aurorarossi) [reply src/alignment_algorithms.jl]: The function that is written in the docstring is not the gradient, but the function from which the gradient is calculated.

File: src/alignment_algorithms.jl (hunk 16 above)
Comment (aurorarossi) [reply src/alignment_algorithms.jl]: I do not know how to improve the number of allocations. Do you have any ideas?

File: src/alignment_algorithms.jl (hunk 22 above)
Comment (aurorarossi) [reply src/alignment_algorithms.jl]: What would be a case of numerical instability here?

File: src/alignment_algorithms.jl (hunk 27 above)
Comment (aurorarossi) [reply src/alignment_algorithms.jl]: Yes, even if we start the algorithm with `α` integer, it will become a float with the iterations.

File: src/alignment_algorithms.jl (hunk 42 above)
Comment (aurorarossi) [reply src/alignment_algorithms.jl]: yes!

File: src/GraphsOptim.jl (hunk 8 above)
Comment (gdalle) [reply src/GraphsOptim.jl:22]: not important for now

File: src/alignment_algorithms.jl (hunk 16 above)
Comment (gdalle) [reply src/alignment_algorithms.jl]: let's keep it that way, it's simpler

File: src/alignment_algorithms.jl (hunk 19 above)
Comment (gdalle) [reply src/alignment_algorithms.jl]: not important

File: src/alignment_algorithms.jl (hunk 22 above)
Comment (gdalle) [reply src/alignment_algorithms.jl]: comparing floating point numbers with zero

File: src/alignment_algorithms.jl (hunk 34 above)
Comment (gdalle) [reply src/alignment_algorithms.jl]: not important

**Review by gdalle — DISMISSED** [review]
Thank you for the contribution @aurorarossi! Let's do the same with #7 now 💪
